packages/simcore-sdk/src/simcore_sdk/node_ports_v2/port.py

```diff
@@ -0,0 +1,123 @@
+import logging
+from pathlib import Path
+from pprint import pformat
+from typing import Any, Dict, Optional, Tuple, Type
+
+from models_library.services import PROPERTY_KEY_RE, ServiceProperty
+from pydantic import Field, PrivateAttr, validator
+
+from ..node_ports.exceptions import InvalidItemTypeError
+from . import port_utils
+from .links import DataItemValue, DownloadLink, FileLink, ItemConcreteValue, PortLink
+
+log = logging.getLogger(__name__)
+
+
+TYPE_TO_PYTYPE: Dict[str, Type[ItemConcreteValue]] = {
+    "integer": int,
+    "number": float,
+    "boolean": bool,
+    "string": str,
+}
+
+
+class Port(ServiceProperty):
+    key: str = Field(..., regex=PROPERTY_KEY_RE)
+    widget: Optional[Dict[str, Any]] = None
+
+    value: Optional[DataItemValue]
+
+    _py_value_type: Tuple[Type[ItemConcreteValue]] = PrivateAttr()
+    _py_value_converter: Type[ItemConcreteValue] = PrivateAttr()
+    _node_ports = PrivateAttr()
+    _used_default_value: bool = PrivateAttr(False)
+
+    @validator("value", always=True)
+    @classmethod
+    def ensure_value(cls, v: DataItemValue, values: Dict[str, Any]) -> DataItemValue:
+        if "property_type" in values and port_utils.is_file_type(
+            values["property_type"]
+        ):
+            if v is not None and not isinstance(v, (FileLink, DownloadLink, PortLink)):
+                raise ValueError(
+                    f"[{values['property_type']}] must follow {FileLink.schema()}, {DownloadLink.schema()} or {PortLink.schema()}"
+                )
+        return v
+
+    def __init__(self, **data: Any):
+        super().__init__(**data)
+
+        self._py_value_type = (
+            (Path, str)
+            if port_utils.is_file_type(self.property_type)
+            else (TYPE_TO_PYTYPE[self.property_type])
+        )
+        self._py_value_converter = (
+            Path
+            if port_utils.is_file_type(self.property_type)
+            else TYPE_TO_PYTYPE[self.property_type]
+        )
+        if (
+            self.value is None
+            and self.default_value is not None
+            and not port_utils.is_file_type(self.property_type)
+        ):
+            self.value = self.default_value
+            self._used_default_value = True
+
+    async def get(self) -> ItemConcreteValue:
+        log.debug(
+            "getting %s[%s] with value %s",
+            self.key,
+            self.property_type,
+            pformat(self.value),
+        )
+
+        if self.value is None:
+            return None
+
+        value = None
+        if isinstance(self.value, PortLink):
```

Comment:
NOTE: next version should have a dispatcher tha registers DataItemValues with each of the functions